cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py

```diff
@@ -0,0 +1,227 @@
+      404: Block not found, OR user lacks permission to edit block.
+      400: Blocks is not linked to upstream content.
+
+Schema for 200 responses, except where noted:
+  {
+      "upstream_ref": string?
+      "version_synced": string?,
+      "version_available": string?,
+      "version_declined": string?,
+      "error_message": string?,
+      "ready_to_sync": Boolean
+  }
+
+Schema for 4XX responses:
+  {
+      "developer_message": string?
+  }
+"""
+from django.contrib.auth.models import AbstractUser
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import UsageKey
+from rest_framework.exceptions import NotFound, ValidationError
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from xblock.core import XBlock
+
+from cms.lib.xblock.upstream_sync import (
+    UpstreamLink, sync_from_upstream, decline_sync, BadUpstream, BadDownstream, fetch_customizable_fields
+)
+from common.djangoapps.student.auth import has_studio_write_access, has_studio_read_access
+from openedx.core.lib.api.view_utils import (
+    DeveloperErrorViewMixin,
+    view_auth_classes,
+)
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.exceptions import ItemNotFoundError
+
+
+# TODO: Potential future view.
+# @view_auth_classes(is_authenticated=True)
+# class DownstreamListView(DeveloperErrorViewMixin, APIView):
+#     """
+#     List all blocks which are linked to upstream content, with optional filtering.
+#     """
+#     def get(self, request: Request) -> Response:
+#         """
+#         Handle the request.
+#         """
+#         course_key_string = request.GET['course_id']
+#         syncable = request.GET['ready_to_sync']
+#         ...
+
+
+@view_auth_classes(is_authenticated=True)
+class DownstreamView(DeveloperErrorViewMixin, APIView):
+    """
+    Inspect or manage an XBlock's link to upstream content.
+    """
+    def get(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Inspect an XBlock's link to upstream content.
+        """
+        assert isinstance(request.user, AbstractUser)
+        downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=False)
+        _ensure_upstream_ref(downstream)
+        if link := UpstreamLink.try_get_for_block(downstream):
+            return Response(link.to_json())
+        raise ValidationError(detail=f"Block '{usage_key_string}' is not linked to an upstream")
+
+    def put(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Edit an XBlock's link to upstream content.
+        """
+        assert isinstance(request.user, AbstractUser)
+        downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=True)
+        new_upstream_ref = request.data.get("upstream_ref")
+        if not isinstance(new_upstream_ref, str):
```

Comment:
I'm not sure I follow. Do you have a more concrete example in mind?

Reply:
I think this is probably my lack of understanding.

Question: In terms of the API and data model, what is the difference between:

1. This course component used to sync to a library, but I want to break that link so it never asks me to sync again, and it actually doesn't even think this was from a library any longer. (Maybe because I never meant for it to be something that syncs.)
2. This course component used to sync to a library, but that library is not available on this site at this time.

Reply:
Gotcha. In the data model:

1. An un-linked block has `block.upstream is None`. So, to sever a block's link, we just set `block.upstream = None`. Calling any of the API methods (`UpstreamLink.get_for_block`, `sync_from_upstream`, `fetch_customizable_fields`) would then return None / be a no-op.
2. If the library is unavailable, `block.upstream` would remain set to the library's key, BUT calling any of the API methods (`UpstreamLink.get_for_block`, `sync_from_upstream`, `fetch_customizable_fields`) will raise a BadUpstream. Calling `UpstreamLink.try_get_for_block` will yield an `UpstreamLink` with a non-null `error_message`.

Does that make sense?

Now, in the HTTP API, I'm realizing that the return codes for each of these scenarios is a bit inconsistent from endpoint to endpoint:

1. unlinked blocks return 400
2. bad links sometimes return 200 with an `error_message`, sometimes return 400

I'm going to think a bit on the HTTP API response codes and push a revision to it tomorrow.

Reply:
Okay @ormsbee , I have pushed a new commit, which changes things a bit.

1. Blocks with no link (or a severed link):
   * In the data model, this is still represented as `block.upstream is None`
     * If the block was previously linked to an upstream, that will be reflected on`block.copied_from_block`.
   * calling any upstream_sync APIs raises NoUpstream
   * calling `GET /api/contentstore/v2/downstreams/{block.usage_key}` yields a 200 with an `error_message` and log entries explaining that the upstream link is missing.
   * calling `POST /api/contentstore/v2/downstreams/{block.usage_key}/sync` yields a 400 with an `error_message` and log entries explaining that the upstream link is missing.

2. Blocks with an malformed link and/or a link to an unavailable upstream:
   * In the data model, `block.upstream` will be set.
   * calling any upstream_sync APIs raises BadUpstream.
   * calling `GET /api/contentstore/v2/downstreams/{block.usage_key}` yields a 200 with an `error_message` and log entries explaining that the upstream link is invalid.
   * calling `POST /api/contentstore/v2/downstreams/{block.usage_key}/sync` yields a 400 with an `error_message` and log entries explaining that the upstream link is missing.

I feel much better about this API, and it ended being net fewer lines of code. Let me know what you think.

Reply:
I think this is great, btw.